x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1484,16 +1485,10 @@ public void onIndexModule(IndexModule module) {
                 assert dlsBitsetCache.get() != null;
                 module.setReaderWrapper(
                     indexService -> new SecurityIndexReaderWrapper(
-                        shardId -> indexService.newSearchExecutionContext(
-                            shardId.id(),
-                            0,
-                            // we pass a null index reader, which is legal and will disable rewrite optimizations
-                            // based on index statistics, which is probably safer...
```

Comment:
^^ This concerns me.
We intentionally passed in a null reader in order to disable rewrite optimizations. I don't think we can blindly undo that.

Reply:
it is more to do with usages of `SearchExecutionContext#getIndexReader` than with `fieldExistsInIndex`. If a searcher is set we may rewrite queries to match none depending on index statistics, for instance for range queries against shards that don't have data within the specified range. That means that certain shards can be skipped entirely. They would not return results if they were to execute the query though, which means that it does not affect the returned results. Unless we rewrite against an empty reader, which is going to yield different results than executing the query against the real reader. I believe here we have the actual reader against the shard available so we should be ok?

Reply:
Thanks for looking into it @javanna! 

>Unless we rewrite against an empty reader, which is going to yield different results than executing the query against the real reader. I believe here we have the actual reader against the shard available so we should be ok?

The reader wrapper is wrapping the actual reader that's used by the `indexService` from what I can see [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/IndexService.java#L268), so it should always be the actual reader against the shard. 

As long as the result of the query doesn't change I think it should be safe from a DLS standpoint. Do you have any additional concerns @tvernum ?